File: private_set_intersection/javascript/src/__tests__/integration.test.ts
import { PSILibrary } from '../implementation/psi'
import PSI from '../wasm_node'

const clientKey = Uint8Array.from([
  0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
  22, 23, 24, 25, 26, 27, 28, 29, 30, 31
])
const serverKey = Uint8Array.from([
  1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22,
  23, 24, 25, 26, 27, 28, 29, 30, 31, 32
])
const fpr = 0.01
const numClientElements = 10
const numServerElements = 100
const clientInputs = Array.from(
  { length: numClientElements },
  (_, i) => `Element ${i}`
)
const serverInputs = Array.from(
  { length: numServerElements },
  (_, i) => `Element ${i * 2}`
)
let psi: PSILibrary

beforeAll(async () => {
  psi = await PSI()
})

describe('PSI Integration', () => {
  test('should return a version string', async () => {
    expect(typeof psi.package.version).toBe('string')
  })

  test('It should create from a static key', async () => {
    const server = psi.server!.createFromKey(serverKey, false)
    expect(server.getPrivateKeyBytes()).toEqual(serverKey)
    const client = psi.client!.createFromKey(clientKey, false)
    expect(client.getPrivateKeyBytes()).toEqual(clientKey)
  })

  test('should compute the intersection', async () => {
    ;[
      { revealIntersection: true, dataStructure: psi.dataStructure.Raw },
      { revealIntersection: true, dataStructure: psi.dataStructure.GCS },
      {
        revealIntersection: true,
        dataStructure: psi.dataStructure.BloomFilter
      },
      { revealIntersection: false, dataStructure: psi.dataStructure.Raw },
      { revealIntersection: false, dataStructure: psi.dataStructure.GCS },
      {
        revealIntersection: false,
        dataStructure: psi.dataStructure.BloomFilter
      }
    ].forEach(({ revealIntersection, dataStructure }) => {
      const client = psi.client!.createWithNewKey(revealIntersection)
      const server = psi.server!.createWithNewKey(revealIntersection)

      const serverSetup = server
        .createSetupMessage(fpr, numClientElements, serverInputs, dataStructure)
        .serializeBinary()

      const clientRequest = client.createRequest(clientInputs).serializeBinary()
      const serverResponse = server
        .processRequest(psi.request.deserializeBinary(clientRequest))
        .serializeBinary()

      if (revealIntersection === true) {
        const intersection = client.getIntersection(
          psi.serverSetup.deserializeBinary(serverSetup),
          psi.response.deserializeBinary(serverResponse)
        )
        const iset = new Set(intersection)
        for (let idx = 0; idx < numClientElements; idx++) {
          if (idx % 2 === 0) {
            // eslint-disable-next-line jest/no-conditional-expect
            expect(iset.has(idx)).toBeTruthy()
          } else {
            // eslint-disable-next-line jest/no-conditional-expect
            expect(iset.has(idx)).toBeFalsy()
          }
          // Test that the intersection is within 10% error margin to match `fpr`
          // eslint-disable-next-line jest/no-conditional-expect
          expect(intersection.length).toBeGreaterThanOrEqual(
            numClientElements / 2
          )
          // eslint-disable-next-line jest/no-conditional-expect
          expect(intersection.length).toBeLessThanOrEqual(
            (numClientElements / 2) * 1.1
          )
        }
      } else {
        const intersection = client.getIntersectionSize(
          psi.serverSetup.deserializeBinary(serverSetup),
          psi.response.deserializeBinary(serverResponse)
        )
        // Test that the intersection is within 10% error margin to match `fpr`
        // eslint-disable-next-line jest/no-conditional-expect
        expect(intersection).toBeGreaterThanOrEqual(
          Math.floor(numClientElements / 2)
        )
        // eslint-disable-next-line jest/no-conditional-expect
        expect(intersection).toBeLessThanOrEqual(
          Math.ceil((numClientElements / 2) * 1.1)
        )
      }
    })
  })
})

File: private_set_intersection/rust/tests/integration_test.rs
use psi::client::PsiClient;
use psi::server::PsiServer;
use psi::*;
use std::collections::HashSet;
use std::iter::FromIterator;

static CLIENT_KEY: &'static [u8] = &[
    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25,
    26, 27, 28, 29, 30, 31,
];
static SERVER_KEY: &'static [u8] = &[
    1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
    27, 28, 29, 30, 31, 32,
];
const FPR: f64 = 0.01;
const NUM_CLIENT_ELEMENTS: usize = 10;
const NUM_SERVER_ELEMENTS: usize = 100;

#[test]
fn test_static_key() {
    let client = PsiClient::create_from_key(CLIENT_KEY, false).unwrap();
    assert_eq!(client.get_private_key_bytes().unwrap(), CLIENT_KEY);

    let server = PsiServer::create_from_key(SERVER_KEY, false).unwrap();
    assert_eq!(server.get_private_key_bytes().unwrap(), SERVER_KEY);
}

#[test]
fn integration_test() {
    for reveal in [false, true] {
        for ds in [
            datastructure::PsiDataStructure::Raw,
            datastructure::PsiDataStructure::Gcs,
            datastructure::PsiDataStructure::BloomFilter,
        ] {
            let client = client::PsiClient::create_with_new_key(reveal).unwrap();
            let server = server::PsiServer::create_with_new_key(reveal).unwrap();

            let client_elements: Vec<String> = (0..NUM_CLIENT_ELEMENTS)
                .map(|i| format!("Element {}", i))
                .collect();
            let server_elements: Vec<String> = (0..NUM_SERVER_ELEMENTS)
                .map(|i| format!("Element {}", i * 2))
                .collect();

            let setup = server
                .create_setup_message(FPR, NUM_CLIENT_ELEMENTS, &server_elements, Some(ds))
                .unwrap();

            // setup message contains server elements
            // setup message is sent to the client

            let request = client.create_request(&client_elements).unwrap();

            // request contains client elements encrypted with client secret key
            // request is sent to the server

            let response = server.process_request(&request).unwrap();

            // response contains client elements encrypted with both client and server secret keys
            // response is sent to the client

            if reveal {
                let intersection = client.get_intersection(&setup, &response).unwrap();
                let set: HashSet<i64> = HashSet::from_iter(intersection.into_iter());

                for i in 0..NUM_CLIENT_ELEMENTS as i64 {
                    // Can fail due to false positives.
                    assert_eq!(set.contains(&i), i % 2 == 0);
                }
            } else {
                let intersection_size = client.get_intersection_size(&setup, &response).unwrap();

                assert!(intersection_size >= (NUM_CLIENT_ELEMENTS / 2));
                assert!(
                    (intersection_size as f64) <= ((NUM_CLIENT_ELEMENTS as f64) / 2.0 * 1.1).ceil()
                );
            }
        }
    }
}

